fe/fe-core/src/main/java/org/apache/doris/datasource/InternalCatalog.java

```diff
@@ -3729,76 +3732,51 @@ public void truncateTable(TruncateTableStmt truncateTableStmt) throws DdlExcepti
         List<Partition> oldPartitions = Lists.newArrayList();
-        Set<Long> oldTabletIds = Sets.newHashSet();
         for (Partition newPartition : newPartitions) {
-            Partition oldPartition = olapTable.replacePartition(newPartition);
+            RecyclePartitionParam recyclePartitionParam = new RecyclePartitionParam();
+            Partition oldPartition = olapTable.replacePartition(newPartition, recyclePartitionParam);
             oldPartitions.add(oldPartition);
-            // save old tablets to be removed
-            for (MaterializedIndex index : oldPartition.getMaterializedIndices(IndexExtState.ALL)) {
-                index.getTablets().forEach(t -> {
-                    oldTabletIds.add(t.getId());
-                });
-            }
+            recyclePartitionParamMap.put(oldPartition.getId(), recyclePartitionParam);
         }
 
         if (isEntireTable) {
             Set<Long> oldPartitionsIds = oldPartitions.stream().map(Partition::getId).collect(Collectors.toSet());
             for (Partition partition : olapTable.getAllTempPartitions()) {
                 if (!oldPartitionsIds.contains(partition.getId())) {
+                    RecyclePartitionParam recyclePartitionParam = new RecyclePartitionParam();
+                    olapTable.fillInfo(partition, recyclePartitionParam);
                     oldPartitions.add(partition);
+                    recyclePartitionParamMap.put(partition.getId(), recyclePartitionParam);
+                    // clear temp partition from memory.
+                    // tablet may be moved to recycle bin or deleted inside
+                    // dropPartitionForTruncate function.
+                    olapTable.dropTempPartition(partition.getName(), false);
                 }
             }
-            // drop all temp partitions
-            olapTable.dropAllTempPartitions();
         }
 
-        // remove the tablets in old partitions
-        for (Long tabletId : oldTabletIds) {
-            Env.getCurrentInvertedIndex().deleteTablet(tabletId);
+        for (Map.Entry<Long, RecyclePartitionParam> pair : recyclePartitionParamMap.entrySet()) {
+            olapTable.dropPartitionForTruncate(olapTable.getDatabase().getId(), isforceDrop, pair.getValue());
         }
 
         return oldPartitions;
     }
 
     public void replayTruncateTable(TruncateTableInfo info) throws MetaNotFoundException {
-        List<Partition> oldPartitions = Lists.newArrayList();
+        boolean isForceDrop = info.getForce();
         Database db = (Database) getDbOrMetaException(info.getDbId());
         OlapTable olapTable = (OlapTable) db.getTableOrMetaException(info.getTblId(), TableType.OLAP);
         olapTable.writeLock();
         try {
-            truncateTableInternal(olapTable, info.getPartitions(), info.isEntireTable());
-
-            // add tablet to inverted index
-            TabletInvertedIndex invertedIndex = Env.getCurrentInvertedIndex();
```

Comment:
this code is to add tablets meta of the new partitions  to TabletInvertIndex.
why delete this code

Reply:
> this code is to add tablets meta of the new partitions to TabletInvertIndex. why delete this code
Thanks.  its adding the tabletInvertIndex for info.getPartitions() , which is old partitions.   i think now it is not necessary to added.

i have tested like,

truncate table, restart FE , and check table content and recover table from recycle bin.
Any thing extra need to test?

Reply:
after truncating table,  restart fe，  have you check TabletInvertIndex contains all new tablets of this Table ?

In my options,  line 3786 ~ 3806 should not delete.  It's used to add new partitions's tablet meta into TabletInvertIndex.

You can see function  InternalCatalog.replayAddPartition,  it's doing the same things,   for replaying the new partition,  it should put its tablet meta into TabletInvertIndex.